File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.java
class Pair{
    int p1;
    int p2;

    Pair(int p1, int p2) {
        this.p1 = p1;
        this.p2 = p2;
    }
};

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        PriorityQueue<Pair> pq = new PriorityQueue<>(
            (ele1, ele2) -> ele2.p2 - ele1.p2
        );
        

        HashMap<Integer,Integer> mp = new HashMap<>();

        for(int i=0;i<nums.length;i++){
            int key = nums[i];
            int val = 1;
            if(mp.containsKey(key) == false){
                mp.put(key,val);
            }else{
                val = mp.get(key);
                mp.put(key,val+1);
            }
        }

        for(Map.Entry<Integer,Integer> ele : mp.entrySet()){
            System.out.println(ele.getKey() +  "-" + ele.getValue());
            Pair p = new Pair(ele.getKey(),ele.getValue());
            pq.add(p);
        }

        int[] ans = new int[k];
        int j = 0;
        while(k > 0){
            ans[j] = pq.poll().p1;
            j++;
            k--;
        }
        return ans;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int countZero = 0;
        int n = nums.size();
        long long productAll = 1;
        vector<int> ans;
        for(auto &num : nums){
            if(num == 0) countZero++;
            else productAll *= num;
        }
        if(countZero == 0){
            vector<int> ans(n, 0);
            for(int i=0;i<n;i++) ans[i] = productAll/nums[i];
            return ans;
        }else if(countZero == 1){
            vector<int> ans(n, 0);
            for(int i=0;i<n;i++) if(nums[i] == 0) ans[i] = productAll;
            return ans;
        }else{
            vector<int> ans(n, 0);
            return ans;
        }
    }
};
